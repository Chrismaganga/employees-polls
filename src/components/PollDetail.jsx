import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

function PollDetail() {
  const { id } = useParams();
  const questions = useSelector(state => state.questions.questions);
  const users = useSelector(state => state.users.users);

  const poll = questions[id];

  if (!poll) {
    return <p>Poll not found</p>;
  }

  const author = users[poll.author];

  return (
    <div className="poll-detail">
      <h2>{author.name} asks:</h2>
      <div className="poll-content">
        <p className="author">{poll.optionOne.text}</p>
        <p className="author">{poll.optionTwo.text}</p>
        <p className="timestamp">Created at: {new Date(poll.timestamp).toLocaleString()}</p>
      </div>
      <button className="vote-button">Vote</button>
    </div>
  );
}

export default PollDetail;
