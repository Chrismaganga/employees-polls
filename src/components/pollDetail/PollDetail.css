.poll-detail {
  max-width: 800px;
  margin: 2rem auto;
  padding: 2rem;
  background-color: #ffffff;
  border-radius: 12px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}

.poll-title {
  text-align: center;
  color: #2c3e50;
  margin-bottom: 2rem;
  font-size: 2rem;
}

.error-message {
  background-color: #fee2e2;
  color: #dc2626;
  padding: 1rem;
  border-radius: 6px;
  margin-bottom: 1rem;
  text-align: center;
}

.options-container {
  display: flex;
  flex-direction: column;
  gap: 2rem;
  margin-bottom: 2rem;
}

.option {
  flex: 1;
  padding: 1.5rem;
  background-color: #f8fafc;
  border-radius: 8px;
  border: 2px solid #e2e8f0;
  transition: all 0.3s ease;
}

.option:hover {
  border-color: #94a3b8;
}

.option.voted {
  border-color: #3b82f6;
  background-color: #eff6ff;
}

.option h3 {
  color: #1e293b;
  margin-bottom: 1rem;
  font-size: 1.25rem;
}

.or-divider {
  text-align: center;
  font-weight: bold;
  color: #64748b;
  margin: 1rem 0;
}

.vote-button {
  width: 100%;
  padding: 0.75rem;
  background-color: #e2e8f0;
  border: none;
  border-radius: 6px;
  color: #475569;
  cursor: pointer;
  transition: all 0.2s ease;
}

.vote-button:hover:not(:disabled) {
  background-color: #cbd5e1;
}

.vote-button.selected {
  background-color: #3b82f6;
  color: white;
}

.vote-button:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.vote-stats {
  margin-top: 1rem;
  position: relative;
}

.vote-bar {
  height: 8px;
  background-color: #3b82f6;
  border-radius: 4px;
  margin-bottom: 0.5rem;
  transition: width 0.3s ease;
}

.vote-stats span {
  color: #64748b;
  font-size: 0.875rem;
}

.submit-vote {
  width: 100%;
  padding: 1rem;
  background-color: #3b82f6;
  color: white;
  border: none;
  border-radius: 8px;
  font-size: 1rem;
  font-weight: 600;
  cursor: pointer;
  transition: all 0.2s ease;
}

.submit-vote:hover:not(:disabled) {
  background-color: #2563eb;
}

.submit-vote:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.total-votes {
  text-align: center;
  color: #64748b;
  margin-top: 1rem;
  font-size: 0.875rem;
}

.poll-detail-error {
  text-align: center;
  color: #dc2626;
  padding: 2rem;
  font-size: 1.25rem;
}

@media (max-width: 768px) {
  .poll-detail {
    padding: 1rem;
    margin: 1rem;
  }

  .options-container {
    flex-direction: column;
    gap: 1rem;
  }

  .or-divider {
    margin: 1rem 0;
  }

  .poll-title {
    font-size: 1.5rem;
  }
}
