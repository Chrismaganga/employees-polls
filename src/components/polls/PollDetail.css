.poll-detail {
  max-width: 800px;
  margin: 2rem auto;
  padding: 2rem;
  background: white;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
}

.poll-title {
  text-align: center;
  color: #333;
  margin-bottom: 2rem;
  font-size: 2rem;
}

.options-container {
  display: flex;
  gap: 2rem;
  margin-bottom: 2rem;
}

.option {
  flex: 1;
  padding: 1.5rem;
  border-radius: 8px;
  background: #f8f9fa;
  transition: transform 0.2s;
}

.option:hover {
  transform: translateY(-2px);
}

.option.voted {
  background: #e3f2fd;
  border: 2px solid #2196f3;
}

.option h3 {
  margin: 0 0 1rem 0;
  color: #333;
  font-size: 1.2rem;
  text-align: center;
}

.or-divider {
  display: flex;
  align-items: center;
  justify-content: center;
  font-weight: bold;
  color: #666;
}

.vote-button {
  width: 100%;
  padding: 0.8rem;
  border: none;
  border-radius: 4px;
  background: #e9ecef;
  color: #495057;
  cursor: pointer;
  transition: background-color 0.2s;
}

.vote-button:hover:not(:disabled) {
  background: #dee2e6;
}

.vote-button.selected {
  background: #007bff;
  color: white;
}

.vote-button:disabled {
  opacity: 0.7;
  cursor: not-allowed;
}

.submit-vote {
  display: block;
  width: 200px;
  margin: 2rem auto;
  padding: 1rem;
  border: none;
  border-radius: 4px;
  background: #28a745;
  color: white;
  font-size: 1rem;
  cursor: pointer;
  transition: background-color 0.2s;
}

.submit-vote:hover:not(:disabled) {
  background: #218838;
}

.submit-vote:disabled {
  opacity: 0.7;
  cursor: not-allowed;
}

.vote-stats {
  margin-top: 1rem;
  position: relative;
}

.vote-bar {
  height: 8px;
  background: #2196f3;
  border-radius: 4px;
  margin-bottom: 0.5rem;
}

.vote-stats span {
  font-size: 0.9rem;
  color: #666;
}

.total-votes {
  text-align: center;
  color: #666;
  margin-top: 1rem;
}

.error-message {
  text-align: center;
  color: #dc3545;
  margin-bottom: 1rem;
}

.poll-detail-error {
  text-align: center;
  color: #dc3545;
  margin: 2rem;
  font-size: 1.2rem;
}

@media (max-width: 768px) {
  .poll-detail {
    padding: 1rem;
    margin: 1rem;
  }

  .options-container {
    flex-direction: column;
    gap: 1rem;
  }

  .or-divider {
    margin: 1rem 0;
  }

  .poll-title {
    font-size: 1.5rem;
  }
}
