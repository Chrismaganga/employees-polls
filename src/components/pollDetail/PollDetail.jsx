import React, { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { answerQuestion } from '../../slices/questionsSlice';
import './PollDetail.css';

function PollDetail() {
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Fetching data from Redux store
  const { questions, status: questionsStatus } = useSelector(state => state.questions);
  const { users } = useSelector(state => state.users);
  const { authedUser, isAuthenticated } = useSelector(state => state.auth);

  const [voteStatus, setVoteStatus] = useState(null);
  const [userVote, setUserVote] = useState(null);
  const [isVoting, setIsVoting] = useState(false);

  // Redirect if not authenticated
  useEffect(() => {
    if (!isAuthenticated) {
      navigate('/login', { state: { from: `/questions/${id}` } });
    }
  }, [isAuthenticated, navigate, id]);

  // Set user's vote if they've already voted
  useEffect(() => {
    if (authedUser && questions?.[id]) {
      const poll = questions[id];
      if (poll.optionOne.votes.includes(authedUser)) {
        setUserVote('optionOne');
      } else if (poll.optionTwo.votes.includes(authedUser)) {
        setUserVote('optionTwo');
      }
    }
  }, [authedUser, questions, id]);

  // Handle loading and error states
  if (questionsStatus === 'loading') {
    return <div className="loading">Loading poll...</div>;
  }

  if (questionsStatus === 'failed') {
    return (
      <div className="error">
        <h2>Error Loading Poll</h2>
        <p>There was a problem loading the poll. Please try again later.</p>
        <button onClick={() => navigate('/')}>Return to Home</button>
      </div>
    );
  }

  const poll = questions?.[id];
  
  // Handle invalid poll ID
  if (!poll && questionsStatus === 'succeeded') {
    return (
      <div className="not-found">
        <h2>404 - Poll Not Found</h2>
        <p>The poll you're looking for doesn't exist.</p>
        <button onClick={() => navigate('/')}>Return to Home</button>
      </div>
    );
  }

  const author = poll ? users?.[poll.author] : null;

  if (!author) {
    return (
      <div className="error">
        <h2>Error Loading Author</h2>
        <p>There was a problem loading the poll author. Please try again later.</p>
        <button onClick={() => navigate('/')}>Return to Home</button>
      </div>
    );
  }

  const handleVote = async (option) => {
    if (!authedUser) {
      navigate('/login', { state: { from: `/questions/${id}` } });
      return;
    }

    if (userVote) {
      setVoteStatus("You've already voted on this poll!");
      return;
    }

    setIsVoting(true);
    setVoteStatus(null);

    try {
      await dispatch(answerQuestion({ authedUser, qid: id, answer: option })).unwrap();
      setUserVote(option);
      setVoteStatus('Your vote has been recorded!');
    } catch (error) {
      setVoteStatus('Error recording your vote. Please try again.');
      console.error('Voting error:', error);
    } finally {
      setIsVoting(false);
    }
  };

  const totalVotes = poll.optionOne.votes.length + poll.optionTwo.votes.length;
  const optionOnePercent = totalVotes > 0 ? (poll.optionOne.votes.length / totalVotes) * 100 : 0;
  const optionTwoPercent = totalVotes > 0 ? (poll.optionTwo.votes.length / totalVotes) * 100 : 0;

  return (
    <div className="poll-detail">
      <h2>{author.name} asks:</h2>
      <div className="poll-content">
        {voteStatus && <p className="vote-status">{voteStatus}</p>}
        
        <div className="poll-option">
          <p className="option-text">{poll.optionOne.text}</p>
          <p className="vote-count">
            {poll.optionOne.votes.length} vote(s) ({optionOnePercent.toFixed(1)}%)
            {userVote === 'optionOne' && <span className="your-vote"> - Your Vote</span>}
          </p>
          {!userVote && (
            <button
              className="vote-button"
              onClick={() => handleVote('optionOne')}
              disabled={isVoting}
            >
              {isVoting ? 'Voting...' : 'Vote for Option One'}
            </button>
          )}
        </div>

        <div className="poll-option">
          <p className="option-text">{poll.optionTwo.text}</p>
          <p className="vote-count">
            {poll.optionTwo.votes.length} vote(s) ({optionTwoPercent.toFixed(1)}%)
            {userVote === 'optionTwo' && <span className="your-vote"> - Your Vote</span>}
          </p>
          {!userVote && (
            <button
              className="vote-button"
              onClick={() => handleVote('optionTwo')}
              disabled={isVoting}
            >
              {isVoting ? 'Voting...' : 'Vote for Option Two'}
            </button>
          )}
        </div>

        <p className="timestamp">Created at: {new Date(poll.timestamp).toLocaleString()}</p>
      </div>
    </div>
  );
}

export default PollDetail;
