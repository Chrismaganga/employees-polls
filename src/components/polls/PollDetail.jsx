import React, { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { startVoting, voteSuccess, voteFailed } from '../../slices/votingSlice';
import './PollDetail.css';

function PollDetail() {
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [selectedOption, setSelectedOption] = useState('');
  
  const { isAuthenticated, userProfile } = useSelector((state) => state.auth);
  const { loading, error } = useSelector(selectVoting);
  const question = useSelector((state) => selectQuestions(state)[id]);
  const votes = useSelector((state) => selectVotes(state, id));

  useEffect(() => {
    if (!isAuthenticated) {
      navigate('/login');
    }
  }, [isAuthenticated, navigate]);

  if (!question) {
    return <div className="poll-detail-error">Poll not found</div>;
  }

  const totalVotes = Object.keys(votes).length;
  const optionOneVotes = Object.values(votes).filter(v => v === 'optionOne').length;
  const optionTwoVotes = Object.values(votes).filter(v => v === 'optionTwo').length;

  const hasVoted = votes[userProfile?.id];

  const calculatePercentage = (optionVotes) => {
    if (totalVotes === 0) return 0;
    return Math.round((optionVotes / totalVotes) * 100);
  };

  const handleVote = async () => {
    if (!selectedOption || !isAuthenticated || hasVoted) return;

    try {
      dispatch(startVoting());
      // Here you would typically make an API call to save the vote
      dispatch(voteSuccess({
        questionId: id,
        option: selectedOption,
        userId: userProfile.id
      }));
    } catch (err) {
      dispatch(voteFailed(err.message));
    }
  };

  return (
    <div className="poll-detail">
      <h2 className="poll-title">Would You Rather...</h2>
      
      {error && <div className="error-message">{error}</div>}
      
      <div className="options-container">
        <div className={`option ${hasVoted && votes[userProfile?.id] === 'optionOne' ? 'voted' : ''}`}>
          <h3>{question.optionOne.text}</h3>
          {hasVoted && (
            <div className="vote-stats">
              <div className="vote-bar" style={{ width: `${calculatePercentage(optionOneVotes)}%` }}></div>
              <span>{optionOneVotes} votes ({calculatePercentage(optionOneVotes)}%)</span>
            </div>
          )}
          {!hasVoted && (
            <button
              className={`vote-button ${selectedOption === 'optionOne' ? 'selected' : ''}`}
              onClick={() => setSelectedOption('optionOne')}
              disabled={loading}
            >
              Select
            </button>
          )}
        </div>

        <div className="or-divider">OR</div>

        <div className={`option ${hasVoted && votes[userProfile?.id] === 'optionTwo' ? 'voted' : ''}`}>
          <h3>{question.optionTwo.text}</h3>
          {hasVoted && (
            <div className="vote-stats">
              <div className="vote-bar" style={{ width: `${calculatePercentage(optionTwoVotes)}%` }}></div>
              <span>{optionTwoVotes} votes ({calculatePercentage(optionTwoVotes)}%)</span>
            </div>
          )}
          {!hasVoted && (
            <button
              className={`vote-button ${selectedOption === 'optionTwo' ? 'selected' : ''}`}
              onClick={() => setSelectedOption('optionTwo')}
              disabled={loading}
            >
              Select
            </button>
          )}
        </div>
      </div>

      {!hasVoted && (
        <button
          className="submit-vote"
          onClick={handleVote}
          disabled={!selectedOption || loading}
        >
          {loading ? 'Submitting...' : 'Submit Vote'}
        </button>
      )}

      <div className="total-votes">
        Total votes: {totalVotes}
      </div>
    </div>
  );
}

export default PollDetail;
