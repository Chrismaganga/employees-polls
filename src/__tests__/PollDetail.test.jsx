import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import '@testing-library/jest-dom';
import PollDetail from '../components/pollDetail/PollDetail';
import { BrowserRouter } from 'react-router-dom';
import votingReducer from '../slices/votingSlice';

const mockPoll = {
  id: 'test123',
  optionOne: { text: 'Test Option One', votes: [] },
  optionTwo: { text: 'Test Option Two', votes: [] },
  author: 'testUser'
};

const mockStore = configureStore({
  reducer: {
    voting: votingReducer
  },
  preloadedState: {
    voting: {
      polls: {
        'test123': mockPoll
      },
      authedUser: 'testUser'
    }
  }
});

// Mock the useParams hook
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ id: 'test123' })
}));

describe('PollDetail Component', () => {
  beforeEach(() => {
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <PollDetail />
        </BrowserRouter>
      </Provider>
    );
    expect(screen.getByText('Test Option One')).toBeInTheDocument();
    expect(screen.getByText('Test Option Two')).toBeInTheDocument();
  });

  it('handles voting when user has voting rights', async () => {
    // Mock the dispatch function
    const mockDispatch = jest.fn(() => Promise.resolve());
    mockStore.dispatch = mockDispatch;

    render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <PollDetail />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.click(screen.getByText('Vote for Option One'));
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('shows error when user has no voting rights', () => {
    const noRightsStore = configureStore({
      reducer: {
        voting: votingReducer
      },
      preloadedState: {
        voting: {
          polls: {
            'test123': {
              ...mockPoll,
              optionOne: { ...mockPoll.optionOne, votes: ['testUser'] }
            }
          },
          authedUser: 'testUser'
        }
      }
    });

    render(
      <Provider store={noRightsStore}>
        <BrowserRouter>
          <PollDetail />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.click(screen.getByText('Vote for Option One'));
    expect(screen.getByText("You don't have voting rights.")).toBeInTheDocument();
  });
});
