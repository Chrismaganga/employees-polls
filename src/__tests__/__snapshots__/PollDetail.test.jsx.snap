// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollDetail Component renders correctly 1`] = `
<div>
  <div
    class="poll-detail"
  >
    <h2>
      Test User
       asks:
    </h2>
    <div
      class="poll-content"
    >
      <p
        class="author"
      >
        Test Option One
      </p>
      <p>
        0
         vote(s) (
        0.0
        %)
      </p>
      <button
        class="vote-button "
      >
        Vote for Option One
      </button>
      <p
        class="author"
      >
        Test Option Two
      </p>
      <p>
        0
         vote(s) (
        0.0
        %)
      </p>
      <button
        class="vote-button "
      >
        Vote for Option Two
      </button>
      <p
        class="timestamp"
      >
        Created at: 
        6/29/2016, 4:21:12 AM
      </p>
    </div>
  </div>
</div>
`;
